flexget/plugins/filter/upgrade.py

```diff
@@ -0,0 +1,170 @@
+    _quality = Column('quality', String)
+    quality = quality_property('_quality')
+    proper = Column(Boolean)
+    added = Column(DateTime, index=True)
+
+    def __init__(self):
+        self.added = datetime.now()
+
+    def __str__(self):
+        return '<Upgrade(id=%s,added=%s,quality=%s)>' % \
+               (self.id, self.added, self.quality)
+
+
+def reject_lower(self, quality, entries):
+    for entry in entries:
+        if quality > entry['quality']:
+            entry.reject('Higher quality already downloaded')
+
+
+def group_entries(entries, identified_by):
+    grouped_entries = defaultdict(list)
+
+    # Group by Identifier
+    for entry in entries:
+        identifier = entry.get('id') if identified_by == 'auto' else entry.render(identified_by)
+        if identifier:
+            grouped_entries[identifier.lower()].append(entry)
+        else:
+            log.debug('No identifier found for', entry['title'])
+
+    return grouped_entries
+
+
+class LazyUpgrade(object):
+    schema = {
+        'oneOf': [
+            {'type': 'boolean'},
+            {
+                'type': 'object',
+                'properties': {
+                    'identified_by': {'type': 'string', 'default': 'auto'},
+                    'tracking': {'type': 'boolean', 'default': True},
+                    'on_lower': {'type': 'string', 'enum': ['accept', 'reject', 'fail', 'skip'], 'default': 'skip'}
+                },
+                'additionalProperties': False
+            }
+        ]
+    }
+
+    def prepare_config(self, config):
+        if not config or isinstance(config, bool):
+            return {'identified_by': 'auto', 'on_lower': 'skip', 'tracking': True}
+        return config
+
+    def on_task_filter(self, task, config):
+        if not config:
+            return
+
+        config = self.prepare_config(config)
+
+        grouped_entries = group_entries(task.entries, config['identified_by'])
+
+        with Session() as session:
+            # Prefetch Data
+            existing_ids = session.query(EntryUpgrade).filter(EntryUpgrade.id.in_(grouped_entries.keys())).all()
+            existing_ids = {e.id: e for e in existing_ids}
+
+            for identifier, entries in grouped_entries.items():
+                if not entries:
+                    continue
+
+                existing = existing_ids.get(identifier)
+
+                if not existing:
+                    # No existing, skip
+                    continue
+
+                # Sort entities in order of quality
+                entries.sort(key=lambda e: e['quality'], reverse=True)
+
+                # First entry will be the best quality
+                if entries[0]['quality'] > existing.quality:
+                    # Accept if we have a better quality
+                    entries[0].accept('upgraded quality')
+
+                # Action lower qualities
+                if config['on_lower'] != 'skip':
+                    entry_actions = {
```

Comment:
Minor thing: You're recreating this dict for every identifier.

Reply:
moved to module level